### PR TITLE
Add rating

### PR DESCRIPTION
**Pull request by DenisZaviryuhin**

### CODE REVIEW

File: src/main/java/com/example/hop_oasis/model/Rating.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+@Setter
+@Entity
+public class Rating {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private Long itemId;
+    @Enumerated(EnumType.STRING)
+    private ItemType itemType;
+    private double rating;
```
Comment (vmakar1134) [inline src/main/java/com/example/hop_oasis/model/Rating.java:19]: double means that you'll have a default 0.0 value for missing values. 
There is a difference in 0.0 and null values for the rating.
0.0 means that rating is very bad
null means that rating is absent

File: src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+public class RatingServiceImpl implements RatingService {
+    private final RatingRepository ratingRepository;
+
+    @Override
+    public double getAverageRating(Long itemId, ItemType itemType){
+        List<Rating> ratings = ratingRepository.findByItemIdAndItemType(itemId, itemType);
+        return ratings.stream()
+                .mapToDouble(Rating::getRating)
+                .average()
+                .orElse(0.0);
```
Comment (vmakar1134) [inline src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java:24]: the same here with default values. 
0.0 means that the item is a very bad one.

File: src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+@Service
+@RequiredArgsConstructor
+public class RatingServiceImpl implements RatingService {
+    private final RatingRepository ratingRepository;
+
+    @Override
+    public double getAverageRating(Long itemId, ItemType itemType){
+        List<Rating> ratings = ratingRepository.findByItemIdAndItemType(itemId, itemType);
+        return ratings.stream()
+                .mapToDouble(Rating::getRating)
```
Comment (vmakar1134) [inline src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java:22]: this is what I said above. 
`Rating::getRating`
this doesn't make sense. Why would  you get rating if you already have ranking?))

File: src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class RatingServiceImpl implements RatingService {
+    private final RatingRepository ratingRepository;
+
+    @Override
+    public double getAverageRating(Long itemId, ItemType itemType){
+        List<Rating> ratings = ratingRepository.findByItemIdAndItemType(itemId, itemType);
```
Comment (vmakar1134) [inline src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java:20]: Can't you calculate avg value in the DB?

File: src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    @Override
+    public int getRatingCount(Long itemId, ItemType itemType){
+       return ratingRepository.countByItemIdAndItemType(itemId, itemType);
+    }
+    @Override
+    public ItemRatingDto getItemRating(Long itemId, ItemType itemType){
+        double averageRating = getAverageRating(itemId, itemType);
+        int ratingCount = getRatingCount(itemId, itemType);
+        return new ItemRatingDto(itemId, itemType, averageRating, ratingCount);
```
Comment (vmakar1134) [inline src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java:34]: this should be done via MapStruct mapper

File: src/main/resources/db/changelog/changes/013-create-rating.sql [hunk 6]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE IF NOT EXISTS rating (
+    id BIGINT AUTO_INCREMENT PRIMARY KEY,
+    item_id BIGINT NOT NULL,
```
Comment (vmakar1134) [inline src/main/resources/db/changelog/changes/013-create-rating.sql:3]: every NOT NULL constraint should be represented in the entity via `@Colum(nullable=false)`
it's just for the other dev to know that this value is not nullable.

File: src/main/java/com/example/hop_oasis/service/data/BeerServiceImpl.java [hunk 7]
```diff
@@ -58,23 +67,50 @@ public Beer save(MultipartFile file, BeerDto beerDto)  {
...
+        return convertToDtoWithRating(beer);
+    }
+
+    private BeerInfoDto convertToDtoWithRating(Beer beer) {
+        BeerInfoDto beerInfoDto = beerInfoMapper.toDto(beer);
+        ItemRatingDto rating = ratingService.getItemRating(beer.getId(), ItemType.BEER);
+        BigDecimal roundedAverageRating = BigDecimal.valueOf(rating.getAverageRating())
+                .setScale(1, RoundingMode.HALF_UP);
+        beerInfoDto.setAverageRating(roundedAverageRating.doubleValue());
+        beerInfoDto.setRatingCount(rating.getRatingCount());
```
Comment (vmakar1134) [inline src/main/java/com/example/hop_oasis/service/data/BeerServiceImpl.java:96]: please, use MapStruct for DTO mapping

**Review by vmakar1134 — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/hop_oasis/controller/BeerController.java [hunk 8]
```diff
@@ -75,6 +75,13 @@ public ResponseEntity<byte[]> addImageToBeer(@RequestParam("beerId") Long beerId
     public ResponseEntity<BeerInfoDto> getBeerById(@PathVariable("id") Long id) {
         return ResponseEntity.ok().body(beerService.getBeerById(id));
     }
+    @PostMapping("/{id}/rating")
```
[the remark this answers is not in the document]
Comment (DenisZaviryuhin) [reply src/main/java/com/example/hop_oasis/controller/BeerController.java:78]: done

File: src/main/java/com/example/hop_oasis/model/Rating.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (DenisZaviryuhin) [reply src/main/java/com/example/hop_oasis/model/Rating.java:19]: done

File: src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java (hunk 3 above)
Comment (DenisZaviryuhin) [reply src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java:22]: done

File: src/main/resources/db/changelog/changes/013-create-rating.sql (hunk 6 above)
Comment (DenisZaviryuhin) [reply src/main/resources/db/changelog/changes/013-create-rating.sql:3]: done

File: src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java (hunk 5 above)
Comment (DenisZaviryuhin) [reply src/main/java/com/example/hop_oasis/service/data/RatingServiceImpl.java:34]: done

File: src/main/java/com/example/hop_oasis/controller/BeerController.java [hunk 9]
```diff
@@ -75,6 +75,13 @@ public ResponseEntity<byte[]> addImageToBeer(@RequestParam("beerId") Long beerId
...
+    public ResponseEntity<BeerInfoDto> addRating(@PathVariable("id") Long id,
+                                          @RequestParam("rating") double rating){
```
[the remark this answers is not in the document]
Comment (DenisZaviryuhin) [reply src/main/java/com/example/hop_oasis/controller/BeerController.java:80]: done

File: src/main/java/com/example/hop_oasis/service/data/BeerServiceImpl.java [hunk 10]
```diff
@@ -58,23 +67,50 @@ public Beer save(MultipartFile file, BeerDto beerDto)  {
...
                 new ResourceNotFoundException(RESOURCE_NOT_FOUND, id));
-        return beerInfoMapper.toDto(beer);
+        return convertToDtoWithRating(beer);
     }
+
+    @Override
+    public BeerInfoDto addRatingAndReturnUpdatedBeerInfo(Long itemId, double ratingValue) {
+        if (ratingValue < 1.0 || ratingValue > 5.0) {
+            throw new IllegalArgumentException("Rating value must be between 1 and 5");
+        }
```
[the remark this answers is not in the document]
Comment (DenisZaviryuhin) [reply src/main/java/com/example/hop_oasis/service/data/BeerServiceImpl.java:74]: done

**Review by vmakar1134 — APPROVED** [review]

**Pull request merged**